### PR TITLE
ENH: show numbers on .info() with verbose flag

### LINKED ISSUE #17304: feature wanted: pd.DataFrame.info() should show line numbers
#### Problem description

When we load data frames, sometimes there are warning of mixed types on some columns via index of columns. It will be then easier for us to check which column is it (by name) if the indices (i.e. line numbers) are shown in pd.DataFrame.info().

### PR DESCRIPTION
**Pull request by Roymprog**
- [x] closes #17304
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by WillAyd** [comment]
This is a duplicate of #28696 which was a continuation of #17304 - did you address comments from those? @Alexandreae were you still working on that?

**Comment by Roymprog** [comment]
Apologies, I completely missed this entire discussion. I was not aware someone else was working on it. I did see the comment left in https://github.com/pandas-dev/pandas/issues/17304, but I started working on it as that was months ago. 

Let me have a look at the comments, then I'll finish up. Still, you can close this if @Alexandreae responds.

**Comment by Roymprog** [comment]
All the suggestions from https://github.com/pandas-dev/pandas/pull/17332 have been implemented. Please let me know what you think

File: doc/source/whatsnew/v0.25.2.rst [hunk 1]
```diff
@@ -97,6 +97,11 @@ Sparse
 
 -
 
+Enhancements
+^^^^^^^^^^^^
+
+- :meth:`df.info()` now shows line numbers for the columns summary (:issue:`17304`)
```
Comment (WillAyd) [inline doc/source/whatsnew/v0.25.2.rst]:
```suggestion
- :meth:`DataFrame.info` now shows line numbers for the columns summary (:issue:`17304`)
```

Can you also move this to v1.0.0?

**Review by WillAyd — CHANGES_REQUESTED** [review]
Cool thanks for picking this up. cc @bashtage who had some thoughts on original design

**Comment by bashtage** [comment]
Output looks pretty good and better than current.  Does it work well with very long variable names? Or other long info properties?  Wondering since the -----  bits are new.

**Comment by Roymprog** [comment]
Sample for long column name variables. 
Script:
```python
col_count = 10
max_col_name = 80
df = pd.DataFrame(np.random.rand(4, col_count), columns=[
          '%s' % (np.random.randint(2, max_col_name)*'a') for x in range(col_count)])
df.info(verbose=True)
```
Output:
```
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4 entries, 0 to 3
Data columns (total 10 columns):
 #   Column                                                                         Non-Null Count  Dtype  
---  ------                                                                         --------------  -----  
 0   aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa                                               4 non-null      float64
 1   aaaaaaaaaaaaaaaaaaaaaaaaaaaa                                                   4 non-null      float64
 2   aaaaaaaa                                                                       4 non-null      float64
 3   aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa  4 non-null      float64
 4   aaaaaaaaaaaaaaaaaaaaaaaaaaaaa                                                  4 non-null      float64
 5   aaaaaaaaaaaaaaaaaaaaaaaaaaa                                                    4 non-null      float64
 6   aaaaaaaaaaaaaaa                                                                4 non-null      float64
 7   aaaaaaaaa                                                                      4 non-null      float64
 8   aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa                     4 non-null      float64
 9   aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa   4 non-null      float64
```
Regarding dtypes, there's a [test](https://github.com/Roymprog/pandas/blob/17304-info-show-numbers/pandas/tests/frame/test_repr_info.py#L288), and for non-null count there is a [docstring test](https://github.com/Roymprog/pandas/blob/17304-info-show-numbers/pandas/core/frame.py#L2379). For the numbering I added an assertion in the [verbosity test](https://github.com/Roymprog/pandas/blob/17304-info-show-numbers/pandas/tests/frame/test_repr_info.py#L209), as I did not feel like posting a thousand line output here.

**Review by jreback — CHANGES_REQUESTED** [review]
this needs some eye balls in it from @pandas-dev/pandas-core

File: doc/source/whatsnew/v0.25.2.rst (hunk 1 above)
Comment (Roymprog) [reply doc/source/whatsnew/v0.25.2.rst]: Moved it to v1.0.0

**Comment by bashtage** [comment]
Look readable in this extreme case.

**Review by WillAyd — APPROVED** [review]
I think this is nice and improves readability

**Comment by WillAyd** [comment]
@pandas-dev/pandas-core anyone else care to take a look?

**Comment by WillAyd** [comment]
ping @pandas-dev/pandas-core for any other comments

File: pandas/core/frame.py [hunk 2]
```diff
@@ -2443,19 +2461,48 @@ def _verbose_repr():
...
+                space_count = max(len_count, max_count) + col_space
+                count_temp = "{count}" + non_null
+            else:
+                count_header = ""
+                space_count = len(count_header)
+                len_count = space_count
+                count_temp = "{count}"
+
+            dtype_header = "Dtype"
+            len_dtype = len("Dtype")
```
Comment (mroeschke) [inline pandas/core/frame.py]: len_dtype = len(dtype_header)

**Review by mroeschke — COMMENTED** [review]
Overall an improvement from the prior output. One small comment.

File: pandas/core/frame.py (hunk 2 above)
Comment (Roymprog) [reply pandas/core/frame.py]: That makes sense, changed it

**Review by mroeschke — APPROVED** [review]

**Comment by WillAyd** [comment]
@jreback care to take a look or merge?

**Comment by jreback** [comment]
yep will again

File: doc/source/whatsnew/v1.0.0.rst [hunk 3]
```diff
@@ -50,6 +50,8 @@ including other versions of pandas.
 Enhancements
 ~~~~~~~~~~~~
 
+- :meth:`Dataframe.info` now shows line numbers for the columns summary (:issue:`17304`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.0.0.rst]: can you make a small sub-section showing these changes. move to the api breaking section as this is a change in the output repr.

File: pandas/core/frame.py [hunk 4]
```diff
@@ -2423,17 +2430,28 @@ def info(
...
+            show_counts = (col_count <= max_cols) and (len(self) < max_rows)
         else:
             show_counts = null_counts
-        exceeds_info_cols = len(self.columns) > max_cols
+        exceeds_info_cols = col_count > max_cols
 
         def _verbose_repr():
             lines.append("Data columns (total %d columns):" % len(self.columns))
-            space = max(len(pprint_thing(k)) for k in self.columns) + 4
+
```
Comment (jreback) [inline pandas/core/frame.py]: can you change this to use .format() (above)

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/frame.py (hunk 4 above)
Comment (Roymprog) [reply pandas/core/frame.py]: Makes sense, added it

File: doc/source/whatsnew/v1.0.0.rst (hunk 3 above)
Comment (Roymprog) [reply doc/source/whatsnew/v1.0.0.rst]: I added the difference to the *Backwards incompatible API changes* section. Hope that is what you meant

**Comment by WillAyd** [comment]
@Roymprog looks like some merge conflicts - can you clean up and re-push?

**Comment by WillAyd** [comment]
@Roymprog can you fix up merge conflict?

**Comment by Roymprog** [comment]
I will in the coming week

**Comment by WillAyd** [comment]
@jreback any outstanding thoughts on this?

File: doc/source/whatsnew/v1.0.0.rst [hunk 5]
```diff
@@ -292,6 +292,34 @@ New repr for :class:`~pandas.arrays.IntervalArray`
 
    pd.arrays.IntervalArray.from_tuples([(0, 1), (2, 3)])
 
+- :meth:`Dataframe.info` now shows line numbers for the columns summary (:issue:`17304`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.0.0.rst]: can you add a title and underline here (like other sub-sections).

**Review by jreback — CHANGES_REQUESTED** [review]
lgtm. small doc-presentation change. ping on green.

File: doc/source/whatsnew/v1.0.0.rst (hunk 5 above)
Comment (Roymprog) [reply doc/source/whatsnew/v1.0.0.rst]: Added an underlined title for the new feature, thanks for the feedback

**Comment by TomAugspurger** [comment]
@Roymprog CI failure:

```
Check for use of not concatenated strings
[error]./pandas/tests/frame/test_repr_info.py:216:String unnecessarily split in two by black. Please merge them manually.
Check for use of not concatenated strings DONE
```

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @Roymprog very . nice!

**Comment by Roymprog** [comment]
Thanks for the fix @TomAugspurger.

Thank you all for all the feedback I received. This was my first open source pull request, thanks for making it happen :smiley:

**Pull request merged**